File: 589-n-ary-tree-preorder-traversal/589-n-ary-tree-preorder-traversal.c
/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     int numChildren;
 *     struct Node** children;
 * };
 */

void treeToarr(struct Node* root,int* arr, int* returnSize){
    if(root != NULL){
        arr[(*returnSize)++] = root->val;
        for(int i = 0; i < root->numChildren; i++){
            treeToarr(root->children[i], arr, returnSize);
        }
    }
}

int* preorder(struct Node* root, int* returnSize) {
    int* arr = malloc(sizeof(int) * 10000);
    *returnSize = 0;
    treeToarr(root, arr, returnSize);
    return arr;
}

File: 589-n-ary-tree-preorder-traversal/589-n-ary-tree-preorder-traversal.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<Integer> preorder(Node root) {
        List<Integer> list = new ArrayList<>();
        preorderTraverse(root, list);
        return list;
    }
    
    public void preorderTraverse(Node root, List<Integer> list) {
        if (root == null)
            return ;
        list.add(root.val);
        int i = 0;
        while (i < root.children.size()){
            preorderTraverse(root.children.get(i), list);
            i++;
        }
    }
}